counsel.el

```diff
@@ -4019,30 +3983,169 @@ TREEP is used to expand internal nodes."
+  the variable `outline-regexp'.
+
+- `:outline-level' is a function of no args to compute the level of
+  an outline heading.  It is called with point at the beginning
+  of `outline-regexp' and with the match data reflicting
+  `outline-regexp'.  The default is to use the value of the
+  variable `outline-level'.
+
+- `:outline-title' is a function of no args to get the title of an
+  outline heading.  It is called with point at the end of
+  `outline-regexp' and with the match data reflicting
+  `outline-regexp'.  The default is to use the function
+  `counsel-outline-title'.
+
+- `:action' is a function of one arg, a marker corresponding to the
+  beginning of the selected outline heading, performing
+  `counsel-outline''s action to jump to this heading.  The
+  default is to use the function `counsel-outline-action'.")
+
+(defun counsel-outline-title ()
+  "Default function used by `counsel-outline' to get the title of
+the current outline heading. See `counsel-outline-settings'."
+  (buffer-substring (point) (line-end-position)))
+
+(defun counsel-outline-title-org ()
+  "Function used by `counsel-outline' to get the title of the
+current outline heading in org-mode buffers. See
+`counsel-outline-settings'."
+  (apply 'org-get-heading (counsel--org-get-heading-args)))
+
+(defun counsel-outline-title-markdown ()
+  "Function used by `counsel-outline' to get the title of the
+current outline heading in markdown-mode buffers. See
+`counsel-outline-title'."
+  ;; `outline-regexp' is set by `markdown-mode' to match both setext
+  ;; (underline) and atx (hash) headings (see
+  ;; `markdown-regex-header').
+  (or (match-string 1) ; setext heading title
+      (match-string 5))) ; atx heading title
+        
+(defun counsel-outline-title-latex ()
+  "Function used by `counsel-outline' to get the title of the
+current outline heading in latex-mode buffers. See
+`counsel-outline-settings'."
+  ;; `outline-regexp' is set by `latex-mode' (see
+  ;; `LaTeX-outline-regexp') to match section macros, in which case we
+  ;; get the section name, as well as `\appendix', `\documentclass',
+  ;; `\begin{document}' and `\end{document}', in which case we simply
+  ;; return that.
+  (if (and (assoc (match-string 1) LaTeX-section-list) ;; section macro
+           (progn ; point is at end of macro name, skip stars and optional args
+             (skip-chars-forward "*")
+             (while (looking-at "\\[")
+               (forward-list))
+             (looking-at "{"))) ; first mandatory arg should be section title
+      (buffer-substring (1+ (point)) (1- (forward-list)))
+    (buffer-substring (line-beginning-position) (point))))
+
+(defun counsel-outline-level-emacs-lisp ()
+  "Function used by `counsel-outline' to compute the level of the
+current outline heading in emacs-lisp-mode buffers. See
+`counsel-outline-settings'."
+  (if (looking-at ";;\\([;*]+\\)")
+      (length (match-string 1))
```

Comment:
This can be simplified as `(- (match-end 1) (match-beginning 1))`, but why is `counsel-outline-level-emacs-lisp` even needed? Can't we reuse `lisp-outline-level`, which is the default buffer-local value of `outline-level` in Lisp modes?

Reply:
Just made the simplification.

`lisp-outline-level` simply computes the length of the `outline-regexp` match, or sets the level to 1000 for the autoload cookie or lines starting with `(`. But to get the correct outine level (1 for `;;;` or `;;*`), we must discard the first two leading `;` as well as any trailing space or tab. (And with the default value of `outline-regexp`, and additional issue would be that the first character of the title is part of the regexp).